### PR TITLE
newpackage: py-awscli (and dependencies)

### PR DESCRIPTION
**Pull request by neilflood**
Package for command line tools for Amazon Web Services. Also 5 other packages it requires which were not already in Spack.  

I have tried to put in all the version restrictions contained in all the various setup.py and requirements.txt files associated with these packages. A few of the single-version requirements seemed overly restrictive, so I have left out a few, but may well be proven wrong in the end. It was a bit of a nightmare to track through......

If anyone thinks the main package should not have the leading "py-" I am open to changing that. I just went with what Spack wanted to call it.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-awscli/package.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+class PyAwscli(PythonPackage):
+    """This package provides a unified command line interface to
+       Amazon Web Services"""
+
+    homepage = "https://pypi.org/project/awscli/"
+    url      = "https://pypi.io/packages/source/a/awscli/awscli-1.16.111.tar.gz"
+
+    version('1.16.111', sha256='63c7900e65a76a1574e1d6e3aaa37c98403339a552428b47559d4e05081f8e4b')
+
+    depends_on('py-setuptools')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-awscli/package.py]: `type='build'`

File: var/spack/repos/builtin/packages/py-awscli/package.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-docutils@0.10:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-awscli/package.py]: Probably `type=('build', 'run')`

File: var/spack/repos/builtin/packages/py-awscli/package.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-nose', type=('build', 'test'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-awscli/package.py]: Probably only `type='test'`

File: var/spack/repos/builtin/packages/py-awscli/package.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-tox@2.3.1:2.999', type='test')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-awscli/package.py]: I don't think this dependency is actually required, probably only needed for version compatibility tests in Travis upstream.

File: var/spack/repos/builtin/packages/py-awscli/package.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-colorama@0.2.5:0.3.9', type=('build', 'run'))
+    depends_on('py-mock@1.3.0:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-awscli/package.py]: Probably only `type='test'`

File: var/spack/repos/builtin/packages/py-awscli/package.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-rsa@3.1.2:3.5.0', type=('build', 'run'))
+    depends_on('py-wheel@0.24.0:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-awscli/package.py]: I don't think this dependency is actually needed to build the package. Probably only if you want to create a wheel.

File: var/spack/repos/builtin/packages/py-awscli/package.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-pyyaml@3.10:3.13', type=('build', 'run'))
+    depends_on('py-argparse', when='^python@2.6:2.6.99', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-awscli/package.py]: `py-argparse@1.1:`

File: var/spack/repos/builtin/packages/py-awscli/package.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    # I suspect there will need to be some version dependencies for these,
+    # too, but they are not explicit. There is an ominous message about tandem
+    # development in the requirements.txt
+    depends_on('py-jmespath', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-awscli/package.py]: This dependency isn't mentioned in `setup.py`, so I'm not sure if it's required.

File: var/spack/repos/builtin/packages/py-awscli/package.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-s3transfer', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-awscli/package.py]: `py-s3transfer@0.2.0:0.2.999'

File: var/spack/repos/builtin/packages/py-awscli/package.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-botocore', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-awscli/package.py]: `py-botocore@1.12.101`
If you don't give it the exact version, setuptools will try to automatically download and install the correct version, which won't work on an air-gapped network.

**Comment by neilflood** [comment]
Thanks @adamjstewart. You have warned me about requirements.txt before, but I never learn. I will write it out 100 times.....

Fixed versions as suggested. It seems you are quite correct that tox is not really necessary, I did not know much about it, but have now removed it. Also realised I had included boto3 without intending to. It was intended to be a separate PR - while it is also a part of Amazon's AWS client code, it is not a requirement for AWS CLI - so I have removed it from this PR.

File: var/spack/repos/builtin/packages/py-botocore/package.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    """A low-level interface to a growing number of Amazon Web Services.
+       The botocore package is the foundation for the AWS CLI as well
+       as boto3"""
+
+    homepage = "https://github.com/boto/botocore"
+    url      = "https://pypi.io/packages/source/b/botocore/botocore-1.12.101.tar.gz"
+
+    version('1.12.101', sha256='46e4daaa7c8cb29237802b63699c16a116f96f301ad2fcfef800574333b58b98')
+    version('1.12.61', sha256='946c24b616cc885d490a6999026125524e85751540eb4af501673d0f5bc7eee1')
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-botocore/package.py]: Missing build dep on setuptools.

File: var/spack/repos/builtin/packages/py-botocore/package.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    depends_on('py-dateutil@2.1:2.7.0', type=('build', 'run'),
+               when='^python@2.6:2.6.999')
+    depends_on('py-dateutil@2.1:2.999', type=('build', 'run'),
+               when='^python@2.7:2.999')
+    depends_on('py-urllib3@1.20:1.25', type=('build', 'run'),
+               when='^python@3.4:')
+    depends_on('py-urllib3@1.20:1.23', type=('build', 'run'),
+               when='^python@3.3')
+    depends_on('py-urllib3@1.20:1.25', type=('build', 'run'),
+               when='^python@2.7')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-botocore/package.py]: Can you confirm that versions like `2.7` actually work? I've had a lot of trouble with shortened versions recently (see #8957)

File: var/spack/repos/builtin/packages/py-jmespath/package.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class PyJmespath(PythonPackage):
+    """JMESPath (pronounced "james path") allows you to declaratively
+       specify how to extract elements from a JSON document"""
+
+    homepage = "git://github.com/boto/jmespath"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-jmespath/package.py]: These links won't open in my browser, you should use the https version instead.

File: var/spack/repos/builtin/packages/py-s3transfer/package.py [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+class PyS3transfer(PythonPackage):
+    """S3transfer is a Python library for managing Amazon S3 transfers"""
+
+    homepage = "git://github.com/boto/s3transfer"
+    url      = "https://pypi.io/packages/source/s/s3transfer/s3transfer-0.2.0.tar.gz"
+
+    version('0.2.0', sha256='f23d5cb7d862b104401d9021fc82e5fa0e0cf57b7660a1331425aab0c691d021')
+    version('0.1.13', sha256='90dc18e028989c609146e241ea153250be451e05ecc0c2832565231dacdf59c1')
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-s3transfer/package.py]: Missing build dep on setuptools

File: var/spack/repos/builtin/packages/py-s3transfer/package.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    depends_on('py-botocore@1.3.0:1.99', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-s3transfer/package.py]: `@1.12.36:1.99`, at least in version 0.2.0

File: var/spack/repos/builtin/packages/py-s3transfer/package.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    depends_on('py-futures@2.2.0:3.999', when='^python@:2.99', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-s3transfer/package.py]: You can just do `@:2` here to capture all of Python 2

File: var/spack/repos/builtin/packages/py-s3transfer/package.py [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+    depends_on('py-nose', type='test')
+    depends_on('py-mock', type='test')
+    depends_on('py-coverage', type='test')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-s3transfer/package.py]: I would just drop this, it's prob not needed for testing, only for coverage reporting.

File: var/spack/repos/builtin/packages/py-botocore/package.py (hunk 12 above)
Comment (neilflood) [reply var/spack/repos/builtin/packages/py-botocore/package.py]: Hmmm... I am not sure if I can confirm it, but it certainly built OK on Python 2.7.15, and selected appropriate versions of these dependencies. However, for safety, I will change them to suitable ranges.

File: var/spack/repos/builtin/packages/py-jmespath/package.py (hunk 13 above)
Comment (neilflood) [reply var/spack/repos/builtin/packages/py-jmespath/package.py]: Goodness, I am not sure where those came from. Yes, https much better.

File: var/spack/repos/builtin/packages/py-s3transfer/package.py (hunk 15 above)
Comment (neilflood) [reply var/spack/repos/builtin/packages/py-s3transfer/package.py]: Apologies. That came from my earlier build, using 0.1.13. Adding explicit variation of the dependency for the two versions listed.

File: var/spack/repos/builtin/packages/py-s3transfer/package.py (hunk 16 above)
Comment (neilflood) [reply var/spack/repos/builtin/packages/py-s3transfer/package.py]: Done

File: var/spack/repos/builtin/packages/py-s3transfer/package.py (hunk 14 above)
Comment (neilflood) [reply var/spack/repos/builtin/packages/py-s3transfer/package.py]: Done

File: var/spack/repos/builtin/packages/py-botocore/package.py (hunk 11 above)
Comment (neilflood) [reply var/spack/repos/builtin/packages/py-botocore/package.py]: Done

File: var/spack/repos/builtin/packages/py-s3transfer/package.py (hunk 17 above)
Comment (neilflood) [reply var/spack/repos/builtin/packages/py-s3transfer/package.py]: OK. It certainly works without it, so done.

**Comment by neilflood** [comment]
@adamjstewart While checking that proj stuff, I noticed that you have been busy adding packages (jmespath, s3transfer and botocore) which I had as part of this PR for awscli. No problem, but I wondered if I should cancel this PR, since you have done most of it separately anyway, and perhaps just start a new one for the awscli package by itself. This PR seems to have been left behind. Let me know what you think.

**Comment by adamjstewart** [comment]
I knew those packages seemed familiar, but I couldn't find them. Sorry about the duplication of effort. Yes, it would probably be easiest to create a new PR with just awscli.

**Comment by neilflood** [comment]
this PR got left behind, and in the meantime all the dependencies were added in by others. Closing this, and will start a new PR just for awscli shortly.

**Pull request closed without merging**